lib/charms/mysql/v0/mysql.py

```diff
@@ -739,7 +739,12 @@ def render_mysqld_configuration(
                 innodb_buffer_pool_chunk_size,
                 group_replication_message_cache_size,
             ) = self.get_innodb_buffer_pool_parameters(available_memory)
-            max_connections = max(self.get_max_connections(available_memory), MIN_MAX_CONNECTIONS)
+
+            # constrain max_connections based on the available memory
+            # after innodb_buffer_pool_size calculation
+            remaining_memory = available_memory - innodb_buffer_pool_size
+            max_connections = max(self.get_max_connections(remaining_memory), MIN_MAX_CONNECTIONS)
```

Comment:
i see the logic behind this, but are we sure we want to limit `max_connections` based on the remaining memory after allocation about 75% to 50% to `innodb_buffer_pool_size`? would it bite us if the max_connections is severly limited compared to client expectations?

Reply:
I agree, and we probably can over commit here. But I'll defer that to @taurus-forever and @delgod

Reply:
Essentially this is (at least) a 3-parameter problem to me. The proposed logic is better than the current one since it can avoid overcommitting memory to 175%. However, keeping limiting user-configurable parameters to 1 doesn't solve the fundamental challenges.

![mysql_major_parameters](https://github.com/canonical/mysql-operator/assets/4356209/25fd50c3-85ac-41ac-9600-b48711333c38)

At this moment, we don't allow configuration other than the parameter "A" in the diagram. I assume that the `mysql.py` logic is shared between the machine charm and the k8s charms so the parameter "A" is determined either by the total amount of physical memory, memory_limit charm config, or pod resource limit. But no other parameters about the memory allocation can be supplied by an user.

When an user needs to bump the max_connections we will have to bump the memory requirement 4 times unnecessarily since there is no way to tell the charm to change the ratio of the memory allocation (75% innodb pool size + 25% max_connections).

For example with MySQL use cases as the backend of OpenStack deployment, we know that it requires a few thousand max connections to cover the distributed (and micro service-ish) control plane services and the multiple workers of each micro service, but 8GB of innodb buffer pool size is more than enough.

Let's say to have max_connections=4,000, 64 GB of the instance is more than enough (4,000 connections * 12MB + 8GB innodb buffer pool size = 56GB). However, with the fixed ratio, we need 200GB machine (4,000 connections * 12MB * (75+25)% / 25% = 192GB).

I cannot think of a better way than moving away from a single parameter.

Reply:
TL;DR: @nobuto-m is correct.

My proposal was a [custom profile tuning](https://docs.google.com/document/d/1pymnut3bB7zhV6hggdNiHAU2VkONZYsoRsqELIzVBkQ/edit) which could be extended to `custom-mem-42%` + `custom-connections-100500` to fine tune the corner case scenarios.

Anyway, by default all pre-allocated memory like `innodb_buffer_pool_size` (and `group_replication_message_cache_size`?) should be excluded from available memory "reserved" for client connections. So this PR still necessary to avoid OOM/crashes.

CC: @7annaba3l , @delgod

Reply:
@nobuto-m is correct.

but before decidion made on independent config options, we need to fix this OOM